File: sources/LiquidSense Release/me/aquavit/liquidsense/ui/client/hud/designer/GuiHudDesigner.java
package me.aquavit.liquidsense.ui.client.hud.designer;

import me.aquavit.liquidsense.LiquidSense;
import me.aquavit.liquidsense.ui.client.hud.element.Element;
import net.minecraft.client.gui.GuiScreen;
import org.lwjgl.input.Keyboard;
import org.lwjgl.input.Mouse;

import java.io.IOException;

public class GuiHudDesigner extends GuiScreen {

    private EditorPanel editorPanel = new EditorPanel(this, 2, 2);
    private Element selectedElement;
    private boolean buttonAction;

    public final Element getSelectedElement() {
        return this.selectedElement;
    }

    public final void setSelectedElement(Element selectedElement) {
        this.selectedElement = selectedElement;
    }

    @Override
    public void initGui() {
        Keyboard.enableRepeatEvents(true);
        this.editorPanel = new EditorPanel(this, this.width / 2, this.height / 2);
        super.initGui();
    }

    @Override
    public void drawScreen(int mouseX, int mouseY, float partialTicks) {
        LiquidSense.hud.render(true);
        LiquidSense.hud.handleMouseMove(mouseX, mouseY);

        if (!LiquidSense.hud.elements.contains(selectedElement))
            selectedElement = null;

        int wheel = Mouse.getDWheel();

        editorPanel.drawPanel(mouseX, mouseY, wheel);

        if (wheel != 0) {
            for (Element element : LiquidSense.hud.elements) {
                if (element.isInBorder(mouseX / element.getScale() - element.getRenderX(),
                        mouseY / element.getScale() - element.getRenderY())) {
                    element.setScale(element.getScale() + (wheel > 0 ? 0.05f : -0.05f));
                    break;
                }
            }
        }
        super.drawScreen(mouseX, mouseY, partialTicks);
    }

    @Override
    protected void mouseClicked(int mouseX, int mouseY, int mouseButton) throws IOException {
        super.mouseClicked(mouseX, mouseY, mouseButton);

        if (buttonAction) {
            buttonAction = false;
            return;
        }

        LiquidSense.hud.handleMouseClick(mouseX, mouseY, mouseButton);

        if (!(mouseX >= editorPanel.x && mouseX <= editorPanel.x + editorPanel.width && mouseY >= editorPanel.y &&
                mouseY <= editorPanel.y + Math.min(editorPanel.realHeight, 200))) {
            selectedElement = null;
            editorPanel.create = false;
        }

        if (mouseButton == 0) {
            for (Element element : LiquidSense.hud.elements) {
                if (element.isInBorder(mouseX / element.getScale() - element.getRenderX(), mouseY / element.getScale() - element.getRenderY())) {
                    this.selectedElement = element;
                    break;
                }
            }
        }
    }

    @Override
    protected void mouseReleased(int mouseX, int mouseY, int state) {
        super.mouseReleased(mouseX, mouseY, state);
        LiquidSense.hud.handleMouseReleased();
    }

    @Override
    public void onGuiClosed() {
        Keyboard.enableRepeatEvents(false);
        LiquidSense.fileManager.saveConfig(LiquidSense.fileManager.hudConfig);
        super.onGuiClosed();
    }

    @Override
    protected void keyTyped(char typedChar, int keyCode) throws IOException {
        switch (keyCode) {
            case Keyboard.KEY_DELETE:{
                if (selectedElement != null) {
                    LiquidSense.hud.removeElement(selectedElement);
                }
                break;
            }
            case Keyboard.KEY_ESCAPE:{
                selectedElement = null;
                editorPanel.create = false;
                break;
            }
            default: {
                LiquidSense.hud.handleKey(typedChar, keyCode);
            }
        }

        super.keyTyped(typedChar, keyCode);
    }
}

File: sources/Kevin Reborn/kevin/hud/designer/GuiHudDesigner.kt
/*
 * This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package kevin.hud.designer

import kevin.hud.element.Element
import kevin.main.KevinClient
import net.minecraft.client.gui.GuiScreen
import org.lwjgl.input.Keyboard
import org.lwjgl.input.Mouse
import kotlin.math.min

class GuiHudDesigner : GuiScreen() {

    private var editorPanel = EditorPanel(this, 2, 2)

    var selectedElement: Element? = null
    private var buttonAction = false

    override fun initGui() {
        Keyboard.enableRepeatEvents(true)
        editorPanel = EditorPanel(this, mc.currentScreen.width / 2, mc.currentScreen.height / 2)
    }

    override fun drawScreen(mouseX: Int, mouseY: Int, partialTicks: Float) {
        KevinClient.hud.render(true)
        KevinClient.hud.handleMouseMove(mouseX, mouseY)

        if (!KevinClient.hud.elements.contains(selectedElement))
            selectedElement = null

        val wheel = Mouse.getDWheel()

        editorPanel.drawPanel(mouseX, mouseY, wheel)

        if (wheel != 0) {
            for (element in KevinClient.hud.elements) {
                if (element.isInBorder(mouseX / element.scale - element.renderX,
                        mouseY / element.scale - element.renderY)) {
                    element.scale = element.scale + if (wheel > 0) 0.05f else -0.05f
                    break
                }
            }
        }
    }

    override fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int) {
        super.mouseClicked(mouseX, mouseY, mouseButton)

        if (buttonAction) {
            buttonAction = false
            return
        }

        KevinClient.hud.handleMouseClick(mouseX, mouseY, mouseButton)

        if (!(mouseX >= editorPanel.x && mouseX <= editorPanel.x + editorPanel.width && mouseY >= editorPanel.y &&
                    mouseY <= editorPanel.y + min(editorPanel.realHeight, 200))) {
            selectedElement = null
            editorPanel.create = false
        }

        if (mouseButton == 0) {
            for (element in  KevinClient.hud.elements) {
                if (element.isInBorder(mouseX / element.scale - element.renderX, mouseY / element.scale - element.renderY)) {
                    selectedElement = element
                    break
                }
            }
        }
    }

    override fun mouseReleased(mouseX: Int, mouseY: Int, state: Int) {
        super.mouseReleased(mouseX, mouseY, state)
        KevinClient.hud.handleMouseReleased()
    }

    override fun onGuiClosed() {
        Keyboard.enableRepeatEvents(false)
        KevinClient.fileManager.saveConfig( KevinClient.fileManager.hudConfig)
        super.onGuiClosed()
    }

    override fun keyTyped(typedChar: Char, keyCode: Int) {
        when (keyCode) {
            Keyboard.KEY_DELETE -> if (Keyboard.KEY_DELETE == keyCode && selectedElement != null)
                KevinClient.hud.removeElement(selectedElement!!)

            Keyboard.KEY_ESCAPE -> {
                selectedElement = null
                editorPanel.create = false
            }

            else ->  KevinClient.hud.handleKey(typedChar, keyCode)
        }

        super.keyTyped(typedChar, keyCode)
    }

    override fun doesGuiPauseGame(): Boolean = false
}